prune/prune.go

```diff
@@ -8,63 +8,87 @@ import (
 
 	"github.com/spacemeshos/go-spacemesh/common/types"
 	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/activesets"
 	"github.com/spacemeshos/go-spacemesh/sql/certificates"
 	"github.com/spacemeshos/go-spacemesh/sql/proposals"
 	"github.com/spacemeshos/go-spacemesh/sql/transactions"
+	"github.com/spacemeshos/go-spacemesh/timesync"
 )
 
-func Prune(
-	ctx context.Context,
-	logger *zap.Logger,
-	db sql.Executor,
-	lc layerClock,
-	safeDist uint32,
-	interval time.Duration,
-) {
-	logger.With().Info("db pruning launched",
-		zap.Uint32("dist", safeDist),
+type Opt func(*Pruner)
+
+func WithLogger(logger *zap.Logger) Opt {
+	return func(p *Pruner) {
+		p.logger = logger
+	}
+}
+
+func New(db *sql.Database, safeDist uint32, activesetEpoch types.EpochID, opts ...Opt) *Pruner {
+	p := &Pruner{
+		logger:         zap.NewNop(),
+		db:             db,
+		safeDist:       safeDist,
+		activesetEpoch: activesetEpoch,
+	}
+	for _, opt := range opts {
+		opt(p)
+	}
+	return p
+}
+
+type Pruner struct {
+	logger         *zap.Logger
+	db             *sql.Database
+	safeDist       uint32
+	activesetEpoch types.EpochID
+}
+
+func Run(ctx context.Context, p *Pruner, clock *timesync.NodeClock, interval time.Duration) {
```

Comment:
Could this be a method of `Pruner`?

Reply:
it can be, i thought that it is better to remove dependencies from Pruner that can be passed only to Run

Reply:
Run takes a pruner as an argument, so someone calling it has to instantiate it anyway?

```go
pruner := prune.New(...)
prune.Run(ctx, pruner, ...)
// vs
pruner := prune.New(...)
pruner.Run(ctx, ...)
```